Change the scheduler plugins FactoryAdapter function to use context parameter to pass logger

mengjiao-liu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
- Change the scheduler plugins new function to pass logger parameter
- Migrated `pkg/scheduler/framework/plugins/nodevolumelimits` to use contextual logging

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1467645065

#### Special notes for your reviewer:
Follow the suggestion of https://github.com/kubernetes/kubernetes/pull/116842#discussion_r1179554937, instead of passing the logger using `handle.Logger()`, we pass the context.
Doing so changes the parameters passed in to initialize New function, and some other plugins parameter will also need to be modified.
```go
// file:  pkg/scheduler/framework/runtime/registry.go

// before
func FactoryAdapter(fts plfeature.Features, withFts PluginFactoryWithFts) PluginFactory {
        return func(plArgs runtime.Object, fh framework.Handle) (framework.Plugin, error) {
                 return withFts(plArgs, fh, fts)

// after
// FactoryAdapter can be used to inject context and feature gates for a plugin that needs
// them when the caller expects the older PluginFactory method.
func FactoryAdapter(ctx context.Context, fts plfeature.Features, withFts PluginFactoryWithFts) PluginFactory {
	return func(plArgs runtime.Object, fh framework.Handle) (framework.Plugin, error) {
		return withFts(ctx, plArgs, fh, fts)
	}
}
```
Some plugins need to pass context during initialization, and the changes are as follows:
```
	registry := runtime.Registry{
                ...
		selectorspread.Name:                  selectorspread.New,
		nodevolumelimits.AzureDiskName:       runtime.FactoryAdapter(ctx, fts, nodevolumelimits.NewAzureDisk),
                ...
	}
```
see https://github.com/kubernetes/kubernetes/commit/ef4ac99f6e027d9ec1b19f7e81df1fd5357a9c94#diff-b5d92dd0897babb368df8c1aad9e95c957eca378bd46481d6d8501011f722715 for more details.

**see https://github.com/kubernetes/kubernetes/pull/116884#discussion_r1315956617 for the newest change.**

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Migrated the nodevolumelimits scheduler plugin to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

/sig instrumentation
/wg structured-logging
/area logging
/sig scheduling

https://github.com/kubernetes/kubernetes/pull/116842#discussion_r1188231165
The plugins(volumebinding #116803 ,examples/stateful https://github.com/kubernetes/kubernetes/pull/116885) can use context after waiting for this PR merge.

mengjiao-liu commented:
/test all

mengjiao-liu commented:
/test all
https://github.com/kubernetes/kubernetes/pull/116842#discussion_r1190184710 Modify PR to pass the context directly.instead of `handle.Logger() `method.

mengjiao-liu commented:
/test all

mengjiao-liu commented:
/remove-sig node
/remove-sig storage

mengjiao-liu commented:
/test pull-kubernetes-verify

mengjiao-liu commented:
/test all

mengjiao-liu commented:
Squash commits when the test is passed.

mengjiao-liu commented:
/hold cancel

mengjiao-liu commented on pkg/scheduler/framework/plugins/noderesources/fit_test.go:
```diff
@@ -496,17 +496,19 @@ func TestEnoughRequests(t *testing.T) {
 			}
 
-			p, err := NewFit(&test.args, nil, plfeature.Features{})
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
```
The `ktesting.NewTestContext` function is not used in this location because of the initialization `NewFit` function of the plugin, the `PreFilter`,`Filter` functions do not call `klog` to output logs.

the same below.

mengjiao-liu replied:
Using the same way is more uniform, so I'll modify it.

logicalhan commented:
/triage accepted
/assign @Richabanker

mengjiao-liu commented:
/cc @alculquicondor

mengjiao-liu commented:
ping @alculquicondor

alculquicondor commented:
/unassign
@kerthcet or @sanposhiho, can you take this?

kerthcet commented:
> /unassign @kerthcet or @sanposhiho, can you take this?

Yes
/assign

mengjiao-liu commented on pkg/scheduler/framework/plugins/schedulinggates/scheduling_gates_test.go:
```diff
@@ -63,12 +63,13 @@ func TestPreEnqueue(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			p, err := New(nil, nil, feature.Features{EnablePodSchedulingReadiness: tt.enablePodSchedulingReadiness})
+			ctx := context.Background()
```
`p.(framework.PreEnqueuePlugin).PreEnqueue(ctx, tt.pod)`  PreEnqueue method has not use the output klog log.

alculquicondor replied:
Use `NewTestContext` for consistency

mengjiao-liu commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -212,7 +214,7 @@ func (pl *CSILimits) filterAttachableVolumes(
 			// If the PVC is invalid, we don't count the volume because
 			// there's no guarantee that it belongs to the running predicate.
-			klog.V(5).InfoS("Unable to look up PVC info", "pod", klog.KObj(pod), "PVC", klog.KRef(pod.Namespace, pvcName))
+			logger.V(5).Info("Unable to look up PVC info", "pod", klog.KObj(pod), "PVC", klog.KRef(pod.Namespace, pvcName))
```
`klog.KObj()` is for an object, but here the `klog.KRef` function takes arguments `pod` and `pvc`(Different object).

Richabanker replied:
ah that's right. sorry about the noise.

mengjiao-liu commented on pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:
```diff
@@ -280,21 +282,21 @@ func (pl *CSILimits) checkAttachableInlineVolume(vol *v1.Volume, csiNode *storag
 	pv, err := pl.pvLister.Get(pvName)
 	if err != nil {
-		klog.V(5).InfoS("Unable to look up PV info for PVC and PV", "PVC", klog.KObj(pvc), "PV", klog.KRef("", pvName))
+		logger.V(5).Info("Unable to look up PV info for PVC and PV", "PVC", klog.KObj(pvc), "PV", klog.KRef("", pvName))
```
see https://github.com/kubernetes/kubernetes/pull/116884/files#r1260697862

mengjiao-liu commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go:
```diff
@@ -336,7 +340,7 @@ func (pl *nonCSILimits) filterVolumes(pod *v1.Pod, newPod bool, filteredVolumes
 			// If the PVC is invalid, we don't count the volume because
 			// there's no guarantee that it belongs to the running predicate.
-			klog.V(4).InfoS("Unable to look up PVC info, assuming PVC doesn't match predicate when counting limits", "pod", klog.KObj(pod), "PVC", klog.KRef(pod.Namespace, pvcName), "err", err)
+			logger.V(4).Info("Unable to look up PVC info, assuming PVC doesn't match predicate when counting limits", "pod", klog.KObj(pod), "PVC", klog.KRef(pod.Namespace, pvcName), "err", err)
```
see https://github.com/kubernetes/kubernetes/pull/116884/files#r1260697862

mengjiao-liu commented on pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go:
```diff
@@ -354,7 +358,7 @@ func (pl *nonCSILimits) filterVolumes(pod *v1.Pod, newPod bool, filteredVolumes
 			// it belongs to the running predicate.
 			if pl.matchProvisioner(pvc) {
-				klog.V(4).InfoS("PVC is not bound, assuming PVC matches predicate when counting limits", "pod", klog.KObj(pod), "PVC", klog.KRef(pod.Namespace, pvcName))
+				logger.V(4).Info("PVC is not bound, assuming PVC matches predicate when counting limits", "pod", klog.KObj(pod), "PVC", klog.KRef(pod.Namespace, pvcName))
```
see https://github.com/kubernetes/kubernetes/pull/116884/files#r1260697862

Richabanker commented:
/lgtm

mengjiao-liu commented:
> Isn't the title of this PR wrong? There are a lot of file changes beyond `nodevolumelimits`.

The purpose is still to migrate this `nodevolumelimits` plugin to use context logging, only if the function used to pass the context parameter does change to many other files.

Of course, we can also change the PR title. Do you have any suggestions?

carlory commented:
cc @sanchayanghosh @codefromthecrypt

kerthcet reviewed (commented):
Only one concern here.

alculquicondor commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -63,7 +63,7 @@ func (pl *DefaultPreemption) Name() string {
 
 // New initializes a new plugin and returns it.
-func New(dpArgs runtime.Object, fh framework.Handle, fts feature.Features) (framework.Plugin, error) {
+func New(_ context.Context, dpArgs runtime.Object, fh framework.Handle, fts feature.Features) (framework.Plugin, error) {
```
since we haven't need a context so far, can we just pass a logger?

mengjiao-liu replied:
Of course, just to be more general, I'll change it to a pass logger.

codefromthecrypt replied:
hmm, I really think it is unwise to assume all use of context is for logger. it is used for tracing, context.Done, all sorts of things. The more things converted from ctx -> logger, more things we have to convert back for extensions.

kerthcet replied:
We initialized with New() for only once here for plugins, so I think it's ok to pass context, but we should use logger when processing scheduling logics. make sense?

alculquicondor replied:
+1
Given the needs for wasm, I'm ok passing a context in this case.

mengjiao-liu commented:
> Let's use NewTestContext uniformly even we don't need the logger, make sense? Same to other places.

Using the same way is more uniform, so I'll modify it.

mengjiao-liu commented:
The CI job `pull-kubernetes-unit` test failed because the latest master branch was not pulled. Updated.

mengjiao-liu commented:
/test pull-kubernetes-kind-dra

kerthcet commented:
/lgtm
/approve
/hold 
for @alculquicondor

alculquicondor commented on pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:
```diff
@@ -51,7 +51,7 @@ var cmpOpts = []cmp.Option{
 
 var (
-	topologySpreadFunc = frameworkruntime.FactoryAdapter(feature.Features{}, New)
+	topologySpreadFunc = frameworkruntime.FactoryAdapter(context.Background(), feature.Features{}, New)
```
When migrating podtopologyspread to contextual logging, we should get rid of the background context.

mengjiao-liu replied:
Sorry for the delay due to K8s code freeze and some other things.

Of course, but in that case, we need to get ctx from the caller (from test) instead.
I talked to pohly about how to deliver ctx in this situation.

>Those must be converted to functions that accept a context. Then when they are called(for example from a test), the caller will be responsible for creating that context.

for example:
before：
```
topologySpreadFunc = frameworkruntime.FactoryAdapter(feature.Features{}, New)
```
after:
```
func topologySpreadFunc(ctx context.Contex) PluginFactory {
    return frameworkruntime.FactoryAdapter(ctx, feature.Features{}, New)
}
```
in tests function, we need to call `topologySpreadFunc(ctx)` to pass ctx.

mengjiao-liu replied:
The latest commit https://github.com/kubernetes/kubernetes/pull/116884/commits/8d94ab5735e6496217b7dabf8b7ed1c054ad7062 to implement the above. Please review it again. @alculquicondor

alculquicondor commented on pkg/scheduler/scheduler_test.go:
```diff
@@ -1009,7 +1009,10 @@ func Test_UnionedGVKs(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
-			registry := plugins.NewInTreeRegistry()
+			_, ctx := ktesting.NewTestContext(t)
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
```
Why add a cancel? Are we spawning any go routine?

mengjiao-liu replied:
Thanks for catch! I will remove the cancel.

alculquicondor replied:
still not removed

alculquicondor reviewed (commented):
/approve

mengjiao-liu commented:
Rebased.

alculquicondor commented on pkg/scheduler/framework/runtime/registry.go:
```diff
@@ -30,13 +31,13 @@ import (
 type PluginFactory = func(configuration runtime.Object, f framework.Handle) (framework.Plugin, error)
```
Also change PluginFactory to accept ctx

mengjiao-liu replied:
Okay. The latest push to implement it.

Commit:
Change the scheduler plugins PluginFactory function to use context parameter to pass logger
- Migrated pkg/scheduler/framework/plugins/nodevolumelimits to use contextual logging
- Fix golangci-lint validation failed
- Check for plugins creation err

mengjiao-liu commented:
/retest

alculquicondor commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -1004,7 +1004,7 @@ func TestRunPreScorePlugins(t *testing.T) {
 				p := p
 				enabled[i].Name = p.name
-				r.Register(p.name, func(_ runtime.Object, fh framework.Handle) (framework.Plugin, error) {
+				r.Register(p.name, func(_ context.Context, _ runtime.Object, fh framework.Handle) (framework.Plugin, error) {
```
check that the error is nil, please.

mengjiao-liu replied:
Done.

mengjiao-liu commented:
Rebased to resolve conflicts.

mengjiao-liu commented:
/retest

mengjiao-liu commented on pkg/scheduler/framework/plugins/nodename/node_name_test.go:
```diff
@@ -55,9 +55,9 @@ func TestNodeName(t *testing.T) {
-			p, _ := New(nil, nil)
-			gotStatus := p.(framework.FilterPlugin).Filter(context.Background(), nil, test.pod, nodeInfo)
+			_, ctx := ktesting.NewTestContext(t)
+			p, _ := New(ctx, nil, nil)
```
Sorry for the delay. I missed the mail notification.

I think this `New` function error always returns nil. Is it necessary to check it?
```
// New initializes a new plugin and returns it.
func New(_ context.Context, _ runtime.Object, _ framework.Handle) (framework.Plugin, error) {
	return &NodeName{}, nil
}
```

alculquicondor replied:
It always returns nil today, but if the implementation changes in the future, we might forget that we are silencing errors here.

mengjiao-liu replied:
I got it. Done.

mengjiao-liu commented on pkg/scheduler/framework/plugins/nodeports/node_ports_test.go:
```diff
@@ -143,9 +143,10 @@ func TestNodePorts(t *testing.T) {
 		t.Run(test.name, func(t *testing.T) {
-			p, _ := New(nil, nil)
+			_, ctx := ktesting.NewTestContext(t)
+			p, _ := New(ctx, nil, nil)
```
Updated.

mengjiao-liu commented on pkg/scheduler/framework/plugins/nodeports/node_ports_test.go:
```diff
@@ -164,13 +165,14 @@ func TestNodePorts(t *testing.T) {
 	node := v1.Node{}
 	nodeInfo.SetNode(&node)
-	p, _ := New(nil, nil)
+	p, _ := New(ctx, nil, nil)
```
Updated.

mengjiao-liu commented on pkg/scheduler/framework/plugins/tainttoleration/taint_toleration_test.go:
```diff
@@ -335,10 +335,11 @@ func TestTaintTolerationFilter(t *testing.T) {
 			nodeInfo.SetNode(test.node)
-			p, _ := New(nil, nil)
-			gotStatus := p.(framework.FilterPlugin).Filter(context.Background(), nil, test.pod, nodeInfo)
+			p, _ := New(ctx, nil, nil)
```
Done.

mengjiao-liu commented:
/test pull-kubernetes-node-e2e-containerd

alculquicondor commented on pkg/scheduler/framework/plugins/nodeunschedulable/node_unschedulable_test.go:
```diff
@@ -75,9 +74,9 @@ func TestNodeUnschedulable(t *testing.T) {
-		p, _ := New(nil, nil)
-		gotStatus := p.(framework.FilterPlugin).Filter(context.Background(), nil, test.pod, nodeInfo)
+		_, ctx := ktesting.NewTestContext(t)
+		p, _ := New(ctx, nil, nil)
```
check the error here too

mengjiao-liu replied:
Thanks for the catch! I checked the files again. Updated.

alculquicondor commented on pkg/scheduler/framework/plugins/tainttoleration/taint_toleration_test.go:
```diff
@@ -237,7 +237,7 @@ func TestTaintTolerationScore(t *testing.T) {
 			fh, _ := runtime.NewFramework(ctx, nil, nil, runtime.WithSnapshotSharedLister(snapshot))
 
-			p, _ := New(nil, fh)
+			p, _ := New(ctx, nil, fh)
```
here as well

mengjiao-liu replied:
Updated.

mengjiao-liu commented:
/hold cancel

mengjiao-liu commented:
/test pull-kubernetes-unit
